thunder/core/utils.py

```diff
@@ -772,7 +772,14 @@ def __repr__(self) -> str:
         return f"{self.__class__.__name__}({{{body}}})"
 
     def __hash__(self) -> int:
-        return hash(frozenset(self.items()))
+        def make_hashable(item) -> tuple:
+            k, v = item
+            if isinstance(v, list):
+                return (k, tuple(v))
+            return item
+
+        hashable_items = map(make_hashable, self.items())
+        return hash(frozenset(hashable_items))
```

Comment:
Items that are put into the FrozenDict should already be hashable similarly to how `frozenset` doesn't accept unhashable types:
```py
In [1]: frozenset(("str", 1, ["list"]))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 frozenset(("str", 1, ["list"]))

TypeError: unhashable type: 'list'
```

Reply:
While this is not the latest commit I do agree with the suggestion of making `FrozenDict` more explicit. It makes sense for it to throw a `TypeError` instead of starting a recursion. I will update such that `make_hashable` is called in the `rhs()` method before creating the `FrozenDict`. This requires to modify the recursive fn too(totally ok), as of now I was banking on the fact that if the dict happens to not have un-hashable values then the `make_hashable` wouldn't have been called.